packages/firestore/src/api/cache_config.ts

```diff
@@ -0,0 +1,152 @@
+/**
+ * @license
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  IndexedDbOfflineComponentProvider,
+  MemoryOfflineComponentProvider,
+  MultiTabOfflineComponentProvider,
+  OfflineComponentProvider,
+  OnlineComponentProvider
+} from '../core/component_provider';
+
+export interface MemoryLocalCache {
+  kind: 'memory';
+  _onlineComponentProvider: OnlineComponentProvider;
+  _offlineComponentProvider: MemoryOfflineComponentProvider;
+}
+
+class MemoryLocalCacheImpl implements MemoryLocalCache {
+  kind: 'memory' = 'memory';
+  _onlineComponentProvider: OnlineComponentProvider;
+  _offlineComponentProvider: MemoryOfflineComponentProvider;
+
+  constructor() {
+    this._onlineComponentProvider = new OnlineComponentProvider();
+    this._offlineComponentProvider = new MemoryOfflineComponentProvider();
+  }
+}
+
+export interface IndexedDbLocalCache {
+  kind: 'indexeddb';
+  _onlineComponentProvider: OnlineComponentProvider;
+  _offlineComponentProvider: OfflineComponentProvider;
+}
+
+class IndexedDbLocalCacheImpl implements IndexedDbLocalCache {
+  kind: 'indexeddb' = 'indexeddb';
+  _onlineComponentProvider: OnlineComponentProvider;
+  _offlineComponentProvider: OfflineComponentProvider;
+
+  constructor(settings: IndexedDbSettings | undefined) {
+    let tabManager: IndexedDbTabManager;
+    if (settings?.tabManager) {
+      settings.tabManager.initialize(settings);
+      tabManager = settings.tabManager;
+    } else {
+      tabManager = indexedDbSingleTabManager(undefined);
+      tabManager.initialize(settings);
+    }
+    this._onlineComponentProvider = tabManager._onlineComponentProvider!;
+    this._offlineComponentProvider = tabManager._offlineComponentProvider!;
+  }
+}
+
+export type FirestoreLocalCache = MemoryLocalCache | IndexedDbLocalCache;
+
+// Factory function
+export function memoryLocalCache(): MemoryLocalCache {
+  return new MemoryLocalCacheImpl();
+}
+
+export interface IndexedDbSettings {
+  cacheSizeBytes?: number;
+  // default to singleTabManager({forceOwnership: false})
+  tabManager?: IndexedDbTabManager;
+}
+
+// Factory function
+export function indexedDbLocalCache(
+  settings?: IndexedDbSettings
+): IndexedDbLocalCache {
+  return new IndexedDbLocalCacheImpl(settings);
+}
+
+export interface IndexedDbSingleTabManager {
+  kind: 'indexedDbSingleTab';
+  initialize: (
+    settings: Omit<IndexedDbSettings, 'tabManager'> | undefined
```

Comment:
I'm curious how using `Omit<...,...>` will turn out in the documentation. IMO, some of our documentation of settings objects can be hard to find or offers limited information. Not sure if this will make the problem worse. We could just define a new settings interface for this case.

Reply:
My intention is to hide this from public API. Renamed it to `_initialize`